src/main/java/htsjdk/samtools/cram/build/ContainerFactory.java

```diff
@@ -52,57 +54,75 @@ public ContainerFactory(final SAMFileHeader samFileHeader, final int recordsPerS
-        final CompressionHeader header = new CompressionHeaderFactory().build(records, substitutionMatrix, coordinateSorted);
+        final CompressionHeader header = new CompressionHeaderFactory().build(records, null, coordinateSorted);
 
         header.readNamesIncluded = preserveReadNames;
 
-        final List<Slice> slices = new ArrayList<Slice>();
+        final List<IndexableSlice> slices = new ArrayList<>();
 
         final Container container = new Container();
         container.header = header;
         container.nofRecords = records.size();
         container.globalRecordCounter = globalRecordCounter;
         container.bases = 0;
         container.blockCount = 0;
+        container.offset = offset;
 
-        long lastGlobalRecordCounter = container.globalRecordCounter;
-        for (int i = 0; i < records.size(); i += recordsPerSlice) {
-            final List<CramCompressionRecord> sliceRecords = records.subList(i,
-                    Math.min(records.size(), i + recordsPerSlice));
-            final Slice slice = buildSlice(sliceRecords, header);
-            slice.globalRecordCounter = lastGlobalRecordCounter;
-            lastGlobalRecordCounter += slice.nofRecords;
-            container.bases += slice.bases;
-            slices.add(slice);
-
-            // assuming one sequence per container max:
-            if (container.sequenceId == -1 && slice.sequenceId != -1)
-                container.sequenceId = slice.sequenceId;
+        // write to a stream to determine byte sizes for Slice BAI Metadata
+        try (final ByteArrayOutputStream osForIndexMetadata = new ByteArrayOutputStream()) {
+            final Version version = CramVersions.CRAM_v3;
+            container.header.write(version, osForIndexMetadata);
+
+            // TODO: is this copy necessary?
+            long globalRecordCounterCopy = container.globalRecordCounter;
+            int sliceIndex = 0;
+            for (int i = 0; i < records.size(); i += recordsPerSlice) {
+                final List<CramCompressionRecord> sliceRecords = records.subList(i,
+                        Math.min(records.size(), i + recordsPerSlice));
+                final Slice slice = buildSlice(sliceRecords, header, globalRecordCounterCopy, refBases);
+                globalRecordCounterCopy += sliceRecords.size();
+
+                // TODO for Container refactoring:
+                // these are also the container.landmarks so let's centralize their construction here
+
+                final int sliceByteOffset = osForIndexMetadata.size();
+                slice.write(version.major, osForIndexMetadata);
+                final int sliceByteSize = osForIndexMetadata.size() - sliceByteOffset;
+
+                slices.add(slice.withIndexingMetadata(sliceByteOffset, sliceByteSize, sliceIndex++));
+
+                container.bases += sliceRecords.stream().map(r -> r.readLength).reduce(Integer::sum).orElse(0);
```

Comment:
`Slice.bases` only existed to supply `Container.bases`, so it's replaced by this step.

Reply:
The old code is:
* Loop through chunks of records.  Populate `Slice.bases` by looping through individual `CramCompressionRecord`s and accumulating their `readLength`s.  Add each Slice's total to `Container.bases`

Is the concern that it's another loop through the record chunk?